library/ajax/track_events.php

```diff
@@ -0,0 +1,136 @@
+<?php
+
+require_once("../../interface/globals.php");
+
+use OpenEMR\Common\Csrf\CsrfUtils;
+use OpenEMR\Common\Uuid\UniqueInstallationUuid;
+use OpenEMR\Services\VersionService;
+
+header("Content-Type: application/json");
+
+// Read and decode the JSON payload.
+$inputJSON = file_get_contents('php://input');
+$data = json_decode($inputJSON, true);
+
+// Verify CSRF token.
+if (!CsrfUtils::verifyCsrfToken($data["csrf_token_form"])) {
+    CsrfUtils::csrfNotVerified();
+    exit;
+}
+if ($data['action'] === 'reportMenuClickData') {
+    reportClicks();
+    //$result = reportUsageData(); // todo: remove this line. for testing until background task is implemented.
+} elseif ($data['action'] === 'reportUsageData') {
+    $result = reportUsageData();
+    echo json_encode($result);
+} else {
+    http_response_code(400);
+    echo json_encode(array("error" => "Invalid action"));
+}
+
+function reportClicks()
+{
+    global $data;
+    $eventType = $data['eventType'] ?? '';
+    $eventLabel = $data['eventLabel'] ?? '';
+    $eventUrl = $data['eventUrl'] ?? '';
```

Comment:
One thing that comes to my mind is if any of the menuURLs would have PHI appended to the query string.  I don't think we do this with any of our menu URLs but I could see someone modifying the standard.json or a module writer adding a url with a specific query string embdded in the click event. 

I wonder if eventUrl should be passed through a URL() object and grab everything but the query params.

There's also the possibility of someone doing a mod_rewrite and putting PHI directly as a directory separator but I think that's more unlikely.  However, if that could be a valid concern perhaps we need to look at having a way to blacklist certain URLs or URL patterns from the background tracking.

Reply:
good point will sanitize